ESQL: GROUP BY ALL

leontyevdv opened the pull request:
The final goal of the task #136253 is to output a list of dimensions for the time-series group. 

We split this work into the three steps:
1. Support bare time-series aggregation functions and output the value of _tsid in the new _timeseries column:
Example: 
```
TS k8s
| STATS avg = avg_over_time(network.cost) BY tbucket = TBUCKET(1 hour)

avg:double         | tbucket:datetime         | _timeseries:keyword
7.262931034482759  | 2024-05-10T00:00:00.000Z | KBGrBhmEnziumfgfOq1dn9NyLSzHbdLj0kfy_m-tXh-yxVR6b3E17ss
6.870689655172414  | 2024-05-10T00:00:00.000Z | KBGrBhmEnziumfgfOq1dn9NyLSzHck3A2MQiFrxrlvTQFP7YxbJ0rYg
6.635416666666667  | 2024-05-10T00:00:00.000Z | KBGrBhmEnziumfgfOq1dn9NyLSzHmDfU38IhLHmDYPRNCcqXvvDolM4
```
2. Load a list of dimensions and output them in the _timeseries column instead of _tsid ([here is a doc about it](https://docs.google.com/document/d/1zyHNj6II0ysp_DwM1ilt7K5XxhlcJgqp4GmkTabx-hk/edit?tab=t.0)). 

3. Make the optimization to load a doc per _tsid.

--- 

This PR adds support for bare time-series aggregate functions and the _timeseries column into the output (**step 1**). 

Part of #136253

not-napoleon committed:
add capability and initial CSV test (WIP)

not-napoleon committed:
tests (that don't pass)

not-napoleon committed:
Merge branch 'main' into esql-group-by-all-over-time

not-napoleon committed:
tests and tracing notes

not-napoleon committed:
Merge branch 'main' into esql-group-by-all-over-time

not-napoleon committed:
added rate agg to tests

not-napoleon committed:
add test gates

not-napoleon committed:
detect when group by all should kick in

not-napoleon committed:
add outline for the rest of the steps

not-napoleon committed:
why does this throw?

not-napoleon committed:
load dimension and metric data in the tests

not-napoleon committed:
build the values aggs for the dimensions

not-napoleon committed:
work around output changed verification

not-napoleon committed:
notes for future me, and others

not-napoleon committed:
draft of analyzer rule

not-napoleon committed:
wire up the rule

not-napoleon committed:
many tests passing

not-napoleon committed:
Merge branch 'main' into esql-group-by-all-over-time

 Conflicts:
	x-pack/plugin/esql/qa/testFixtures/src/main/resources/k8s-timeseries.csv-spec

not-napoleon committed:
failing for a new reason

Commit:
[CI] Auto commit changes from spotless

not-napoleon committed:
misc cleanups

not-napoleon committed:
Merge branch 'main' into esql-group-by-all-over-time

 Conflicts:
	x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/LoadMapping.java
	x-pack/plugin/esql/qa/testFixtures/src/main/resources/k8s-timeseries.csv-spec

not-napoleon committed:
more testing

not-napoleon committed:
Merge remote-tracking branch 'refs/remotes/not-napoleon/esql-group-by-all-over-time' into esql-group-by-all-over-time

not-napoleon committed:
forgot to add this test yesterday

Commit:
[CI] Auto commit changes from spotless

Commit:
[CI] Update transport version definitions

leontyevdv committed:
Merge branch 'main' into esql-group-by-all-over-time

# Conflicts:
#	server/src/main/resources/transport/upper_bounds/8.18.csv
#	server/src/main/resources/transport/upper_bounds/8.19.csv
#	server/src/main/resources/transport/upper_bounds/9.0.csv
#	server/src/main/resources/transport/upper_bounds/9.1.csv
#	server/src/main/resources/transport/upper_bounds/9.2.csv
#	server/src/main/resources/transport/upper_bounds/9.3.csv
#	x-pack/plugin/esql/qa/testFixtures/src/main/resources/k8s-timeseries.csv-spec
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/TranslateTimeSeriesAggregate.java
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Drop.java

leontyevdv committed:
Merge branch 'main' into esql-group-by-all-over-time

leontyevdv committed:
Fix unit tests

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all

leontyevdv committed:
ESQL: Add GROUP BY ALL

Group by _tsid and add it to the output.

Closes #136253

leontyevdv committed:
ESQL: Add GROUP BY ALL

Group by _tsid and add it to the output.

Closes #136253

dnhatn commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -526,20 +526,63 @@ private Aggregate resolveAggregate(Aggregate aggregate, List<Attribute> children
+            LogicalPlan newChild = aggregate.child();
+            if (aggregate instanceof TimeSeriesAggregate) {
+                for (Expression g : groupings) {
+                    if (g instanceof UnresolvedAttribute ua && ua.name().equals(MetadataAttribute.TSID_FIELD)) {
```
Do we still need the change here?

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all

leontyevdv committed:
ESQL: Add GROUP BY ALL

Group by _tsid and add it to the output in the _timeseries metadata
attribute.

Closes #136253

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all

leontyevdv committed:
ESQL: Add GROUP BY ALL

Group by _tsid and add it to the output in the _timeseries metadata
attribute.

Part of #136253

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all

# Conflicts:
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/TranslateTimeSeriesAggregate.java

leontyevdv committed:
ESQL: Add GROUP BY ALL

Group by _tsid and add it to the output in the _timeseries metadata
attribute.

Part of #136253

leontyevdv committed:
ESQL: Add GROUP BY ALL

Fix tests

Part of #136253

leontyevdv committed:
Update docs/changelog/137367.yaml

leontyevdv committed:
ESQL: Add GROUP BY ALL

Fix tests

Part of #136253

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all

leontyevdv committed:
ESQL: Add GROUP BY ALL

Cleanup comments

Part of #136253

leontyevdv committed:
Merge remote-tracking branch 'origin/feature/esql-group-by-all' into feature/esql-group-by-all

leontyevdv committed:
ESQL: Add GROUP BY ALL

Cleanup

Part of #136253

leontyevdv committed:
ESQL: Add GROUP BY ALL

Cleanup

Part of #136253

leontyevdv committed:
ESQL: Add GROUP BY ALL

Cleanup

Part of #136253

leontyevdv committed:
ESQL: Add GROUP BY ALL

Ignore GroupByAll in GenerativeForkRestTest

Part of #136253

leontyevdv committed:
Update docs/changelog/137367.yaml

dnhatn reviewed (commented):
The approach looks good. Thanks Dima! I played with the PR and think we can simplify the translation and dimension loading. Can you take a look at this [patch](https://github.com/user-attachments/files/23639573/group-all.patch) to see if I missed anything?

leontyevdv committed:
ESQL: Add GROUP BY ALL

Remove Drop JavaDoc

Part of #136253

leontyevdv committed:
ESQL: Add GROUP BY ALL

Address the PR's comments

Part of #136253

leontyevdv committed:
ESQL: Add GROUP BY ALL

Address the PR's comments

Part of #136253

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all

# Conflicts:
#	x-pack/plugin/esql/qa/testFixtures/src/main/resources/k8s-timeseries.csv-spec

leontyevdv committed:
ESQL: Add GROUP BY ALL

Address the PR's comments

Part of #136253

leontyevdv committed:
ESQL: Add GROUP BY ALL

Address the PR's comments

Part of #136253

leontyevdv commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/TimeSeriesGroupByAll.java:
```diff
@@ -0,0 +1,90 @@
+                    newAggregateFunctions.add(new Alias(alias.source(), alias.name(), new Values(tsAgg.source(), tsAgg)));
+                } else {
+                    // Preserve non-time-series aggregates
+                    newAggregateFunctions.add(agg);
```
I added a check for this below. If there are mixed aggs, we throw an exception now.

leontyevdv committed:
ESQL: Add GROUP BY ALL

Address the PR's comments

Part of #136253

leontyevdv commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/PostOptimizationPhasePlanVerifier.java:
```diff
@@ -85,7 +86,15 @@ private static void verifyOutputNotChanged(QueryPlan<?> optimizedPlan, List<Attr
             );
-            boolean ignoreError = hasProjectAwayColumns || hasLookupJoinExec || hasTextGroupingInTimeSeries;
+
+            // TranslateTimeSeriesAggregate may add a _timeseries attribute into the projection
```
We need this because in the LogicalPlanOptimizer.optimize() [line 123](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java#L124C9-L124C43) we take an expected output from the verified plan and the _timeseries attribute is not there. It appears in the optimized plan.

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all

# Conflicts:
#	x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/generative/GenerativeForkRestTest.java

dnhatn commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/k8s-timeseries.csv-spec:
```diff
@@ -784,5 +784,184 @@ max_deriv:double | max_rate:double | time_bucket:datetime     | cluster:keyword
+
+count:long | _timeseries:keyword
+29         | KBGrBhmEnziumfgfOq1dn9NyLSzHck3A2MQiFrxrlvTQFP7YxbJ0rYg=
+29         | KBGrBhmEnziumfgfOq1dn9NyLSzHbdLj0kfy/m+tXh+yxVR6b3E17ss=
```
Maybe add sorting by _timeseries, since the same counts can appear in any order.

dnhatn reviewed (approved):
Looks great. Thanks Dima!

felixbarny commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/TimeSeriesGroupByAll.java:
```diff
@@ -0,0 +1,100 @@
+        );
+        List<Expression> groupings = new ArrayList<>();
+        groupings.add(timeSeries);
+        groupings.addAll(aggregate.groupings());
```
My 2c is that this is a bit confusing. It suggests that we're just grouping by cluster but there's no outer aggregation for it. I understand that it groups by all and then also adds the cluster column but I don't think that's self-evident.

Maybe it's something we should leave out for now, it's something we can easily add later as well. Maybe it makes more sense in combination with an explicit `BY DIMENSIONS(), cluster`. But I'm skeptical if it's an important use case to support both an implicit and explicit grouping at the same time.

leontyevdv replied:
I added a check that rejects such cases. It allows to use GroupingFunction (Bucket, TBucket, Categorize) but rejects attributes. This applies to the time-series aggregates only.

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all

leontyevdv committed:
ESQL: Add GROUP BY ALL

Address the PR's comments

Part of #136253

leontyevdv committed:
ESQL: Add GROUP BY ALL

Remove support of the grouping attributes for the bare time-series
aggregates.

Part of #136253

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all

leontyevdv committed:
ESQL: Add GROUP BY ALL

Fix merge.

Part of #136253

leontyevdv committed:
Merge branch 'main' into feature/esql-group-by-all
